src/rulesets/LinearAlgebra/structured.jl

```diff
@@ -19,6 +19,13 @@ function rrule(::typeof(diag), A::AbstractMatrix)
     return diag(A), diag_pullback
 end
 
+function rrule(::typeof(*), D::Diagonal{<:Real}, V::AbstractVector{<:Real})
+    function times_pullback(Ȳ)
+        return (NO_FIELDS, @thunk(Diagonal(Ȳ .* V)), @thunk(D.diag .* Ȳ))
```

Comment:
```suggestion
        return (NO_FIELDS, @thunk(Diagonal(Ȳ .* V)), @thunk(D * Ȳ))
```
My benchmarks show that  `D * Ȳ` is just as performant as  `D.diag .* Ȳ`
and I think it is clearer
```
julia> dd*k == dd.diag .* k
true

julia> @show typeof(dd); @show size(dd)
typeof(dd) = Diagonal{Float64,Array{Float64,1}}
size(dd) = (10000, 10000)

julia> @show typeof(k); @show size(k);
typeof(k) = Array{Float64,1}
size(k) = (10000,)


julia> @btime $dd.diag .* $k;
  5.250 μs (2 allocations: 78.20 KiB)

julia> @btime $dd * $k;
  5.257 μs (2 allocations: 78.20 KiB)
```